Source/WebKit/UIProcess/API/Cocoa/WKDownload.mm

```diff
@@ -202,14 +202,54 @@ void processDidCrash(WebKit::DownloadProxy& download) final
         [m_delegate download:wrapper(download) didFailWithError:[NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorNetworkConnectionLost userInfo:nil] resumeData:nil];
     }
 
+#if HAVE(MODERN_DOWNLOADPROGRESS)
+    void didReceivePlaceholderURL(WebKit::DownloadProxy& download, const WTF::URL& url, std::span<const uint8_t> bookmarkData, CompletionHandler<void()>&& completionHandler) final
+    {
+        if (!m_delegate || !m_respondsToDidReceivePlaceholderURL) {
+            completionHandler();
+            return;
+        }
+
+        BOOL bookmarkDataIsStale = NO;
+        NSError *bookmarkResolvingError;
+        RetainPtr data = toNSData(bookmarkData);
+        RetainPtr urlFromBookmark = adoptNS([[NSURL alloc] initByResolvingBookmarkData:data.get() options:0 relativeToURL:nil bookmarkDataIsStale:&bookmarkDataIsStale error:&bookmarkResolvingError]);
+        if (bookmarkResolvingError || bookmarkDataIsStale)
+            RELEASE_LOG_ERROR(Network, "Failed to resolve URL from bookmark data");
+
+        NSURL *placeholderURL = urlFromBookmark ? urlFromBookmark.get() : (NSURL *)url;
+
+        [m_delegate _download:wrapper(download) didReceivePlaceholderURL:placeholderURL completionHandler:makeBlockPtr(WTFMove(completionHandler)).get()];
+    }
+
+    void didReceiveFinalURL(WebKit::DownloadProxy& download, const WTF::URL& url, std::span<const uint8_t> bookmarkData) final
+    {
+        if (!m_delegate || !m_respondsToDidReceiveFinalURL)
+            return;
+
+        BOOL bookmarkDataIsStale = NO;
+        NSError *bookmarkResolvingError;
+        RetainPtr data = toNSData(bookmarkData);
+        RetainPtr urlFromBookmark = adoptNS([[NSURL alloc] initByResolvingBookmarkData:data.get() options:0 relativeToURL:nil bookmarkDataIsStale:&bookmarkDataIsStale error:&bookmarkResolvingError]);
+        if (bookmarkResolvingError || bookmarkDataIsStale)
+            RELEASE_LOG_ERROR(Network, "Failed to resolve URL from bookmark data");
+
+        NSURL *finalURL = urlFromBookmark ? urlFromBookmark.get() : (NSURL *)url;
```

Comment:
I think ObjC can use the ?: operator, and I think the explicit cast to (NSURL *) might not be necessary.